Feature/financial years

BerendWouters opened the pull request:
I'm still in doubt about the domain model of the FinancialYear.

I've added some unit tests nonetheless, so we can still change the implementation.

vyruz1986 commented on Persistence/Repositories/AttachmentStorage.cs:
```diff
@@ -50,3 +48,11 @@ public void Delete(string blobUri, CancellationToken cancellationToken)
     }
 }
+
+public interface IAttachmentStorage
```
This interface should remain in the Domain project. See https://stackoverflow.com/a/72399927 for reasons why

vyruz1986 commented on Persistence/Repositories/BankAccountRepository.cs:
```diff
@@ -41,3 +40,12 @@ public async Task SaveAsync(CancellationToken ct)
     }
 }
+
+public interface IBankAccountRepository
```
This interface should remain in the Domain project. See https://stackoverflow.com/a/72399927 for reasons why

vyruz1986 commented on Persistence/Repositories/FinancialYearRepository.cs:
```diff
@@ -0,0 +1,39 @@
+    }
+}
+
+public interface IFinancialYearRepository
```
This interface should be in the Domain project. See https://stackoverflow.com/a/72399927 for reasons why

vyruz1986 commented on Persistence/Repositories/MemberRepository.cs:
```diff
@@ -45,3 +44,13 @@ public async Task Save(CancellationToken cancellationToken)
 
 }
+
+public interface IMemberRepository
```
This interface should remain in the Domain project. See https://stackoverflow.com/a/72399927 for reasons why

vyruz1986 commented on Persistence/HaSpManContext.cs:
```diff
@@ -22,9 +22,16 @@ public HaSpManContext(DbContextOptions<HaSpManContext> options)
     protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
-            => optionsBuilder.UseSqlServer(x => x.MigrationsHistoryTable("__EFMigrationsHistory", Schema.HaSpMan));
+    {
+        if (!optionsBuilder.IsConfigured)
```
Why do we need this?

BerendWouters replied:
For unit testing purposes, we can't have the provider already configured.

vyruz1986 reviewed (changes requested):
I'm also missing an EFCore migration. I'm also wondering whether FInancialYear should become an AggregateRoot whjch owns many Transaction objects. The it is now, they would become 2 aggregate roots with relations between them, which is generally frowned upon in DDD.
A better would be to make FinanalYear the aggregate root, and make the current Transaction aggregate root, a value object of a financial year instead. In this case the TransactionRepository would need to be removed, and each interaction with a transaction would need to take please through the FinancialYear. 
In this an EFCore configuration using `OwnsMany(y => y.Transactions)` would be needed from the FinancialYear config

BerendWouters commented on Commands/Handlers/FinancialYear/AddFinancialYear/AddFinancialYearCommand.cs:
```diff
@@ -0,0 +1,14 @@
+{
+    public AddFinancialYearCommandValidator()
+    {
+        RuleFor(x => x.StartDate).NotEmpty();
```
Ah yes, read over the part of the year being a configurable thing.

BerendWouters commented on Commands/Handlers/FinancialYear/AddFinancialYear/AddFinancialYearHandler.cs:
```diff
@@ -0,0 +1,36 @@
+    }
+    public async Task<Guid> Handle(AddFinancialYearCommand request, CancellationToken cancellationToken)
+    {
+        await EnsureStartDateNotInExistingFinancialYear(request, cancellationToken);
```
Correct, will update once I've figured out the way to handle configurable start of a new financial year.

BerendWouters commented:
> I'm also missing an EFCore migration. I'm also wondering whether FInancialYear should become an AggregateRoot whjch owns many Transaction objects. The it is now, they would become 2 aggregate roots with relations between them, which is generally frowned upon in DDD. A better would be to make FinanalYear the aggregate root, and make the current Transaction aggregate root, a value object of a financial year instead. In this case the TransactionRepository would need to be removed, and each interaction with a transaction would need to take please through the FinancialYear. In this an EFCore configuration using `OwnsMany(y => y.Transactions)` would be needed from the FinancialYear config
I was opting for that last path too, but I did want to have this discussed. From a domain point of view, this approach makes to most sense. So: FinancialYear becomes the new aggregate root, Transaction is removed and becomes a value object of that newly created aggregate root.
I do wonder about the following though:
The existing transaction handlers are unaware of the Financial year concept. Am I correct in letting them exist, but instead of interacting with the Transaction (which it can't since we removed the ITransactionRepository), it should first figure out what FinancialYear the transaction belongs to, using the `ReceivedDate` of the transaction? This would imply that we could move transactions from one FinancialYear to another (as longs as they're not closed and so on).
The cool thing about using the FinancialYear as the aggregate root is that we only have to check if the transaction can be edited if the FinancialYear isn't closed, and not the transaction themselves. As a consequence, I don't think we'd need to add a `Locked` property on the Transaction.

The reason I didnt yet create a migration was to prevent too many up and down migrations during development.

vyruz1986 commented:
> > I'm also missing an EFCore migration. I'm also wondering whether FInancialYear should become an AggregateRoot whjch owns many Transaction objects. The it is now, they would become 2 aggregate roots with relations between them, which is generally frowned upon in DDD. A better would be to make FinanalYear the aggregate root, and make the current Transaction aggregate root, a value object of a financial year instead. In this case the TransactionRepository would need to be removed, and each interaction with a transaction would need to take please through the FinancialYear. In this an EFCore configuration using `OwnsMany(y => y.Transactions)` would be needed from the FinancialYear config
> 
> I was opting for that last path too, but I did want to have this discussed. From a domain point of view, this approach makes to most sense. So: FinancialYear becomes the new aggregate root, Transaction is removed and becomes a value object of that newly created aggregate root. I do wonder about the following though: The existing transaction handlers are unaware of the Financial year concept. Am I correct in letting them exist, but instead of interacting with the Transaction (which it can't since we removed the ITransactionRepository), it should first figure out what FinancialYear the transaction belongs to, using the `ReceivedDate` of the transaction? This would imply that we could move transactions from one FinancialYear to another (as longs as they're not closed and so on). The cool thing about using the FinancialYear as the aggregate root is that we only have to check if the transaction can be edited if the FinancialYear isn't closed, and not the transaction themselves. As a consequence, I don't think we'd need to add a `Locked` property on the Transaction.
> 
> The reason I didnt yet create a migration was to prevent too many up and down migrations during development.

Indeed that would impose a breaking change in that transactions could not be accessed by themselves, but would need to be accessed through their FY aggregate root.
Which means that would likely translate back to UI changes as well, as you would first need to select the financial year before being able to see/add/edit transactions...

BerendWouters commented:
I've updated the PR.

No migrations yet, but I think the domain change has been done.

However, some other issues I'm not sure if the current implementation is the best:
- As you mentioned, handling transactions should be done through the aggregate root. I've removed the TransactionRepository, but allowed to find the FY via a transactionId. I'm not really sure if this goes against any practices. It feels off, but there's a difference between the UI and the actual domain model (which I think _is_ allowed in DDD).
- However, by introducing the FY as the aggregate root and DDD stating you should always start at the aggregate root, we will always track all transactions (and attachments) under one FY. This _might_ introduce a performance penalty, but unsure. It's not quite clear to me to what DDD says about including a certain child item, but it feels like a bad practice.
- Querying. DDD is aimed at managing the state of the object you're working on, but is if I remember correctly quite loose about getting data. For that reason, we still have a DbSet of Transactions, but only for querying purposes. I think this is fine, but we should always adhere the no-tracking configuration, when querying the DbContext. In the past I've used a query facade pattern to make this more strict, but that's just convention imho.

vyruz1986 commented:
> * Querying. DDD is aimed at managing the state of the object you're working on, but is if I remember correctly quite loose about getting data. For that reason, we still have a DbSet of Transactions, but only for querying purposes. I think this is fine, but we should always adhere the no-tracking configuration, when querying the DbContext. In the past I've used a query facade pattern to make this more strict, but that's just convention imho.

I think the DbSet will have to go anyway, because iirc an entity cannot have both it's own DbSet as well as be owned by another entity.
Offcourse you could expose an `IQueryable` from the DbContext which does something like `FinancialYears.SelectMany(y => t>Transactions).AsNoTracking()`, or go all the way as you mentioned and create an `I<Entity>QueryStore` next to `I<Entity>Repository` for query purposes

BerendWouters commented:
I knew this would happen: owned types don't support hierarchy (https://learn.microsoft.com/en-us/ef/core/modeling/owned-entities#current-shortcomings).

vyruz1986 commented on Domain/FinancialYearConfiguration.cs:
```diff
@@ -0,0 +1,20 @@
+﻿namespace Domain;
+
+public class FinancialYearConfiguration
```
This shouldn't be an aggregate root I think? It should just be configuration which is retrieved wherever needed using the .NET `IOptions` pattern

vyruz1986 commented on Domain/Transaction.cs:
```diff
@@ -131,6 +131,7 @@ public void AddAttachments(ICollection<TransactionAttachment> attachments)
             Attachments.Add(attachment);
         }
     }
+    
```
?

vyruz1986 commented on Persistence/EntityConfigurations/FinancialYearEntityTypeConfiguration.cs:
```diff
@@ -0,0 +1,19 @@
+        builder.Property(x => x.EndDate).IsRequired();
+        builder.Property(x => x.IsClosed).IsRequired();
+        // Owned entity types cannot have inheritance hierarchies https://learn.microsoft.com/en-us/ef/core/modeling/owned-entities#current-shortcomings
+        builder.HasMany(x => x.Transactions);
```
This should be `OwnsMany()` I think?

BerendWouters replied:
Yes, but OwnsMany doesn't support inheritance of collections (yet): https://learn.microsoft.com/en-us/ef/core/modeling/owned-entities#current-shortcomings

vyruz1986 commented on Persistence/Migrations/20230626190616_AddFinancialYearAsAggregateRoot.cs:
```diff
@@ -0,0 +1,92 @@
+                schema: "HaSpMan",
+                table: "Transactions",
+                type: "uniqueidentifier",
+                nullable: true);
```
This shouldn't be nullable eventually, but of course it must be temporarily nullable for all existing transactions. However the should be some migration SQL code in here to add the column (with nullable: true), update the value for all existing transactions, then make it non-nullable for all new transactions

vyruz1986 commented on Persistence/EntityConfigurations/TransactionConfigurations.cs:
```diff
@@ -5,6 +5,18 @@
 
 namespace Persistence.EntityConfigurations;
 
+public class TransactionConfiguration : IEntityTypeConfiguration<Transaction>
```
This should be an Action in the `.OwnsMany()` I commented above

vyruz1986 commented on Persistence/EntityConfigurations/FinancialYearEntityTypeConfiguration.cs:
```diff
@@ -0,0 +1,19 @@
+{
+    public void Configure(EntityTypeBuilder<FinancialYear> builder)
+    {
+        builder.Property(x => x.StartDate).IsRequired();
+        builder.Property(x => x.EndDate).IsRequired();
+        builder.Property(x => x.IsClosed).IsRequired();
```
These shouldn't be required

vyruz1986 commented on Persistence/Migrations/20230728184205_TransactionsBelongToFinancialYear.cs:
```diff
@@ -0,0 +1,36 @@
+                             "HaspMan.FinancialYears (Id, StartDate, EndDate, IsClosed) " +
+                             "VALUES (newId(), CAST('2022-09-01 00:00:00.0000000 +02:00' AS DATETIMEOFFSET),CAST('2023-08-31 00:00:00.0000000 +02:00' AS DATETIMEOFFSET), 0)");
+
+        migrationBuilder.DropIndex("IX_Transactions_FinancialYearId", schema: "HaspMan", table: "Transactions");
+        migrationBuilder.AlterColumn<Guid>(
+            "FinancialYearId", 
+            schema:"HaspMan",
+            table: "Transactions",
+            nullable: false);
+        migrationBuilder.CreateIndex(
+            name: "IX_Transactions_FinancialYearId",
+            schema: "HaSpMan",
+            table: "Transactions",
+            column: "FinancialYearId");
```
This only sets the field to non-nullable in the index, however the table should be altered as well
